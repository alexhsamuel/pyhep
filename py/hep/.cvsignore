*.pyc
*.pyo
config.py
*-install
pdt.pickle
