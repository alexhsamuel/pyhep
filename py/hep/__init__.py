#-----------------------------------------------------------------------
#
# module hep
#
# Copyright 2003 by Alex Samuel.  All rights reserved.
#
#-----------------------------------------------------------------------

"""Module for general-purpose code."""

#-----------------------------------------------------------------------
# imports
#-----------------------------------------------------------------------

from __future__ import generators

import os
import signal
import traceback

#-----------------------------------------------------------------------
# exceptions
#-----------------------------------------------------------------------

class TimeoutException(Exception):
    """A timeout occurred."""

    pass



class NotSupportedError(RuntimeError):
    """A feature is not supported."""

    pass



class SyntaxError(Exception):

    pass



#-----------------------------------------------------------------------
# classes
#-----------------------------------------------------------------------

class Token:

    pass



class Timeout:

    def __timeout_handler(self, signal_number, frame):
        assert signal_number == signal.SIGALRM
        signal.signal(signal.SIGALRM, self.__old_handler)
        self.__active = 0
        raise TimeoutException, self.__message


    def __init__(self, time, message):
        self.__message = message
        self.__old_handler = \
            signal.signal(signal.SIGALRM, self.__timeout_handler)
        self.__active = 1
        signal.alarm(time)


    def cancel(self):
        if not self.__active:
            return
        signal.alarm(0)
        signal.signal(signal.SIGALRM, signal.SIG_DFL)
        self.__active = 0



class MappingAsObject:
    """Implement the object attribute protocol using a mapping.

    A 'MappingAsObject' wraps an object satifying the mapping protocol.
    Attribute references are converted to item gets and sets in the
    underlying mapping."""

    def __init__(self, mapping):
        """Wrap 'mapping' in the object procotol."""
        
        self.__dict__["_MappingAsObject__mapping"] = mapping


    _mapping = property(lambda self: self.__mapping)
    """Read-only access to the underlying mapping."""


    def __getattr__(self, name):
        try:
            return self.__mapping[name]
        except KeyError:
            raise AttributeError, name


    def __setattr__(self, name, value):
        try:
            self.__mapping[name] = value
        except KeyError, exception:
            raise AttributeError, str(exception)



class NrBreitWigner:
    """A non-resonant Breit-Wigner line shape amplitude factor."""

    def __init__(self, mass, width):
        """Construct a non-relativistic Breit-Wigner function.

        'mass' -- The resonance mass.
        
        'width' -- The resonance width."""

        self.__mass = mass
        self.__width = width


    def __call__(self, s):
        """Return the amplitude at energy-squared 's'."""

        mass = self.__mass
        return 1 / (s - complex(mass * mass, mass * self.__width))



#-----------------------------------------------------------------------
# functions
#-----------------------------------------------------------------------

try:
    enumerate = __builtins__.enumerate
except AttributeError:
    def enumerate(iterable):
        count = 0
        for item in iter(iterable):
            yield count, item
            count += 1


try:
    sorted = __builtins__.sorted
except AttributeError:
    def sorted(iterable, cmp=cmp):
        result = list(iterable)
        result.sort(cmp)
        return result


try:
    reversed = __builtins__.reversed
except AttributeError:
    def reversed(iterable):
        result = list(iterable)
        result.reverse()
        return result
    

def sort(sequence, sort_fn=None):
    """Return a copy of 'sequence', sorted by 'sort_fn'.

    'sort_fn' -- The comparison function to use.  If 'None', the default
    (built-in 'cmp') is used."""

    sequence = list(sequence)
    if sort_fn is None:
        sort_fn = cmp
    sequence.sort(sort_fn)
    return sequence


def binarySearch(sequence, value, compare=cmp):
    """Find 'value' in sorted 'sequence' by binary search.

    'sequence' -- A sequence of values.  It is assumed to be sorted by
    'compare'.

    'compare' -- A comparison function.

    returns -- The index of the largest element in 'sequence' smaller
    than 'value', or -1 if there is no such element."""

    if len(sequence) == 0:
        return -1
    if compare(value, sequence[0]) < 0:
        return -1
    if compare(value, sequence[-1]) > 0:
        return len(squence) - 1

    i0 = 0
    i2 = len(sequence) - 1
    while True:
        if i2 <= i0 + 1:
            return i2 - 1
        i1 = (i0 + i2) // 2
        if compare(value, sequence[i1]) < 0:
            i2 = i1
        else:
            i0 = i1


def mapFromKeys(key_sequence, value_function):
    """Construct a map from a sequence of keys.

    A map is created whose keys are the elements of 'key_sequence'.  For
    each key, the corresponding value is constructed from the key using
    'value_function'.  If 'sequence' contains a particular key more than
    once, only one is placed in the map (which one is undefined).

    'key_sequence' -- A sequence of key values.

    'value_function' -- A callable for determining the value for each
    element of the sequence.  It is invoked for each key, passing the
    kay as the single argument."""
    
    result = {}
    for key in key_sequence:
        result[key] = value_function(key)
    return result


def mapFromValues(key_function, value_sequence):
    """Construct a map from a sequence of values.

    A map is created whose values are the elements of 'value_sequence'.
    For each value, the corresponding key is constructed from the value
    using 'key_function'.  If 'key_function' assigns the same key to
    more than one element of the sequence, only one element is actually
    used in the map (which one is undefined).

    'key_function' -- A callable for determining the key for each
    element of the sequence.  It is invoked for each value, passing
    the value as the single argument.

    'value_sequence' -- A sequence of values."""
    
    result = {}
    for value in value_sequence:
        result[key_function(value)] = value
    return result


def invertMap(map):
    """Produce a dictionary by inverting keys and values of 'map'."""

    return dict([ (value, key) for key, value in map.items() ])


def remove(dict, key):
    """Get and remove a dictionary item.

    'dict' -- The dictionary from which to get and remove the item.

    'key' -- The key of the item to get and remove.

    returns -- The value corresponding to 'key' in 'dict'.  

    raises -- 'KeyError' if 'dict' has no key 'key'."""
    
    value = dict[key]
    del dict[key]
    return value


def popkey(map, key, default=None):
    """Get and remove a dictionary item.

    'dict' -- The dictionary from which to get and remove the item.

    'key' -- The key of the item to get and remove.

    'default' -- The value to return if 'key' is not in 'map'.

    returns -- The value corresponding to 'key' in 'dict', or 'default'
    if none."""

    if key in map:
        result = map[key]
        del map[key]
        return result
    else:
        return default


def getHostName():
    """Return the name of this host."""

    import socket

    try:
        return socket.gethostbyname_ex(socket.gethostname())[0]
    except socket.error:
        return "(unknown)"


def isMap(object):
    """Determine whether an object satisfies the mapping protocol.

    returns -- True if 'object' satisfies the map protocol as
    specified.""" 

    # FIXME.  There probably is a more consistent way to do this.

    # If the object is an instance of the standard 'dict' class (or a
    # subclass of it), it must satisfy the mapping protocol.
    if isinstance(object, dict):
        return True
    # Check for the required methods.
    for method_name in ("get", "__getitem__", "__len__", ):
        if not hasattr(object, method_name):
            return False
    # The object checks out.
    return 1


def mkdirRecursively(path):
    """Create a directory, including parent directories if necessary."""

    parent, name = os.path.split(path)
    if parent != "" and not os.path.isdir(parent):
        mkdirRecursively(parent)
    os.mkdir(path)


def rmRecursively(path):
    """Remove the file or directory at 'path'.

    If 'path' is a directory, its contents are removed first."""

    if os.path.isdir(path):
        # Construct a list of contents of the directory.
        contents = map(lambda e, path=path: os.path.join(path, e),
                       os.listdir(path))
        # Remove them all.
        map(rmRecursively, contents)
        # Remove the directory itself.
        os.rmdir(path)
    else:
        os.unlink(path)


