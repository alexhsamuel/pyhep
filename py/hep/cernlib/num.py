#-----------------------------------------------------------------------
#
# num.py
#
# Copyright (C) 2004 by Alex Samuel.  All rights reserved.
#
#-----------------------------------------------------------------------

"""Numerical functions based on CERNLIB."""

#-----------------------------------------------------------------------
# imports
#-----------------------------------------------------------------------

import ext

#-----------------------------------------------------------------------
# functions
#-----------------------------------------------------------------------

def chiSquareCDF(x, n):
    """CDF for chi-square distribution.

    'n' -- The chi-squared parameter; must be one or larger.

    returns -- The probability that a random variable chosen
    according to a chi-square distribution with 'n' degrees of
    freedom will have a value of 'x' or lower."""

    return 1 - ext.prob(x, n)


