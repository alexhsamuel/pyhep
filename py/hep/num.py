#--------------------------------------------------*- coding: Latin1 -*-
#
# num.py
#
#-----------------------------------------------------------------------

"""Numerical and mathematical functions and code."""

#-----------------------------------------------------------------------
# imports
#-----------------------------------------------------------------------

from __future__ import division
from __future__ import generators

# from   ext import erf, erfc
import math
import operator
import random

#-----------------------------------------------------------------------
# constants
#-----------------------------------------------------------------------

golden_ratio = (1 + math.sqrt(5)) / 2

#-----------------------------------------------------------------------
# classes
#-----------------------------------------------------------------------

class Statistic(float):
    """A number with statistical uncertainty.

    A 'Statistic' object is a 'float' with an extra attribute
    'uncertainty' that contains a statistical uncertainty.  The
    statistic can be used in ordinary arithmetical operations, and the
    uncertainty is propagated assuming each statistic is from an
    independent normal distribution.

    Create a statistic by specifying its value and uncertainty,
    e.g. 'Statistic(val, unc)'.  Both the value and uncertainty are
    converted to 'float'.  You may also specify the value and
    uncertainty together in a string of the form '"val +- unc"' or
    '"val ± unc"'.  If the uncertainty is unspecified or 'None', the
    square root of the value is used."""

    def __new__(class_, value, uncertainty=None):
        # Convert moments to statistics.
        if isinstance(value, Moments):
            val = value.mean
            unc = value.standard_deviation
        # Handle strings specially.
        elif isinstance(value, str):
            # If the uncertainty was specified in the string, extract it.
            if "±" in value:
                val, unc = map(float, value.split("±", 1))
            elif value.find("+-") != -1:
                val, unc = map(float, value.split("+-", 1))
            else:
                # No uncertainty in the string.  Treat the string as the
                # value only.
                val = float(value)
                unc = None
        else:
            # Convert the value to a float value.
            val = float(value)
            # Take the uncertainty from the 'uncertainty' attribute, if
            # present.
            if hasattr(value, "uncertainty"):
                unc = value.uncertainty
            else:
                unc = None

        # The 'uncertainty' argument overrides uncertainty extracted
        # from 'value'.
        if uncertainty is not None:
            unc = float(uncertainty)
        # Still no uncertainty?  Use the square root of the value.
        if unc is None:
            unc = math.sqrt(abs(val))

        # Make sure the uncertainty is positive.
        if unc < 0:
            raise ValueError, "negative uncertainty %f" % unc

        # Build the statistic.
        statistic = float.__new__(class_, val)
        statistic.__uncertainty = unc
        return statistic


    uncertainty = property(lambda self: self.__uncertainty)


    def __repr__(self):
        return "Statistic(%f, %f)" % (float(self), self.uncertainty)


    def __str__(self):
        return "%f ± %f" % (float(self), self.uncertainty)


    def __neg__(self):
        return Statistic(-float(self), self.uncertainty)


    def __add__(self, other):
        if hasattr(other, "uncertainty"):
            return Statistic(
                float(self) + float(other),
                math.hypot(self.uncertainty, other.uncertainty))
        else:
            return Statistic(float(self) + other, self.uncertainty)
        

    def __radd__(self, other):
        return Statistic.__add__(self, other)

        
    def __sub__(self, other):
        if hasattr(other, "uncertainty"):
            return Statistic(
                float(self) - float(other),
                math.hypot(self.uncertainty, other.uncertainty))
        else:
            return Statistic(float(self) - other, self.uncertainty)


    def __rsub__(self, other):
        return -Statistic.__sub__(self, other)


    def __mul__(self, other):
        if hasattr(other, "uncertainty"):
            return Statistic(
                float(self) * float(other),
                math.hypot(float(self) * other.uncertainty,
                           float(other) * self.uncertainty))
        else:
            return Statistic(float(self) * other, self.uncertainty * other)
    

    def __rmul__(self, other):
        return Statistic.__mul__(self, other)


    def __div__(self, other):
        if hasattr(other, "uncertainty"):
            return Statistic(
                float(self) / float(other),
                math.hypot(float(self) * other.uncertainty,
                           float(other) * self.uncertainty)
                / float(other) ** 2)
        else:
            return Statistic(float(self) / other, self.uncertainty / other)


    def __rdiv__(self, other):
        if hasattr(other, "uncertainty"):
            return Statistic(
                float(other) / float(self),
                math.hypot(float(self) * other.uncertainty,
                           float(other) * self.uncertainty)
                / float(self) ** 2)
        else:
            return Statistic(other / float(self), self.uncertainty / other)


    # Use the same division operators for true division.
    __truediv__ = __div__
    __rtruediv__ = __rdiv__


    def format(self, value_digits=6, uncertainty_digits=None, exponent=None):
        if uncertainty_digits is None:
            uncertainty_digits = value_digits
        if exponent is None:
            return "%.*f ± %.*f" \
                   % (value_digits, float(self),
                      uncertainty_digits, self.uncertainty)
        elif exponent == "%":
            if value_digits > 0:
                value_size = 4 + value_digits
            else:
                value_size = 3
            if uncertainty_digits > 0:
                uncertainty_size = 3 + uncertainty_digits
            else:
                uncertainty_size = 2
            return "(%*.*f ± %*.*f)%%" \
                   % (value_size, value_digits,
                      100 * float(self),
                      uncertainty_size, uncertainty_digits,
                      100 * self.uncertainty)
        elif isinstance(exponent, int):
            scale = 10 ** (-exponent)
            return "(%.*f ± %.*f)x10^%d" \
                   % (value_digits, scale * float(self),
                      uncertainty_digits, scale * self.uncertainty,
                      exponent)



#-----------------------------------------------------------------------

class Efficiency:
    """The efficiency of a sampled process.

    Estimates the binonmial parameter from a sampling of trials from a
    binomial process.  The trials may be weighted.

    The 'float' value of an efficiency object is 'p / t', where 'p' is
    the number of passing samples (sum of weights of passing samples)
    and 't' is the total number of samples (sum of weights of all
    samples).  The 'uncertainty' attribute holds an estimate of the
    uncertainty on the efficiency, 'p (t - p) / t ** 3'."""


    def __init__(self):
        self.tries = 0
        self.passes = 0.0


    def __repr__(self):
        return "Efficiency(tries=%r, passes=%r)" \
               % (self.tries, self.passes)


    def __str__(self):
        return self.format()


    def sample(self, outcome, weight=1):
        """Record the outcome of a trial.

        'outcome' -- A boolean value, indicating whether the trial was
        successful.

        'weight' -- The statistical weight of this sample."""
        
        weight = float(weight)
        self.tries += weight
        if outcome:
            self.passes += weight


    def __lshift__(self, outcome):
        """Equivalent to 'sample' with unit weight."""

        self.sample(outcome)


    def __float__(self):
        if self.tries != 0:
            return self.passes / self.tries
        else:
            return 0


    def __get_uncertainty(self):
        p = self.passes
        t = self.tries
        if t != 0:
            return math.sqrt(p * (t - p) / (t ** 3))
        else:
            return 0


    uncertainty = property(__get_uncertainty)


    def format(self, digits=1, exponent="%"):
        return Statistic(self).format(digits, digits, exponent)
        


#-----------------------------------------------------------------------

class Moments(object):

    def __init__(self, numbers=()):
        self.count = 0
        self.sum = 0
        self.sum_sq = 0
        for number in numbers:
            self << number


    def __lshift__(self, number):
        self.count += 1
        self.sum += number
        self.sum_sq += number * number


    def __get_mean(self):
        n = self.count
        if n < 1:
            return 0
        else:
            return self.sum / n


    mean = property(__get_mean)


    def __get_variance(self):
        n = self.count
        if n < 2:
            return 0
        return n / (n - 1) * (self.sum_sq / n - (self.sum / n) ** 2)


    variance = property(__get_variance)


    def __get_standard_deviation(self):
        # Due to roundoff errors, the variance may be slightly
        # negative.  Return zero in that case.
        return math.sqrt(max(self.variance, 0))


    standard_deviation = property(__get_standard_deviation)
            

    def __get_statistic(self):
        return Statistic(self.mean, self.standard_deviation)


    statistic = property(__get_statistic)



#-----------------------------------------------------------------------
# functions
#-----------------------------------------------------------------------

def binary(x, bits=32):
    """Represent the least significant 'bits' of 'x' in binary form."""

    result = ""
    x = long(x)
    bit = 1l << (bits - 1)
    while bit > 0:
        if x & bit != 0:
            result += "1"
        else:
            result += "0"
        bit >>= 1
    return result


def gaussian(mu, sigma, x):
    """Return a probability density from a Gaussian distribution.

    'mu' -- The distribution's mean.

    'sigma' -- The distribution's standard deviation.

    'x' -- The value at which to evaluate the probability density."""

    return 0.3989422804014326779399 \
           * math.exp(-0.5 * ((x - mu) / sigma)**2) \
           / sigma


def almost_equal(num0, num1, fraction=1e-5):
    """Return true if 'num0' and 'num1' are almost equal.

    'fraction' -- The fraction of the larger of 'num0' and 'num1'
    within which the values are almost equal."""

    return abs(num0 - num1) < fraction * max(abs(num0), abs(num1))


def get_bit(value, bit):
    """Return 'True' if bit 'bit' is set in 'value'."""
    
    return value & (1 << bit) != 0


def hypot(*terms):
    """Generalized multidimensional hypotenuse.

    'terms' -- Any number of numerical terms.

    returns -- The square root of the sum of squares of 'terms'."""
    
    return math.sqrt(
        reduce(lambda sum, term: sum + term * term,
               terms,
               0))


def mhypot(first_term, *terms):
    """Generalized multidimensional hypotenuse with Minkowski metric.

    returns -- The square root of the square of the 'first_term' minus
    the sum of squares of remaining '*terms'."""
    
    return math.sqrt(reduce(lambda sum, term: sum - term * term,
                            terms, first_term * first_term))


def factorial(n):
    """Compute the factorial of 'n'."""

    # Check the argument.
    if n < 0:
        raise ValueError, "'n' must be nonnegative"
    if n != long(n):
        raise ValueError, "'n' must be integral"
    # Compute the factorial.
    # FIXME: Use a less naive algorithm.
    result = 1
    for i in xrange(2, n + 1):
        result *= i
    return result


def in_range(min_value, value, max_value):
    """Return true if 'value' is in a semi-open interval."""

    return value >= min_value and value < max_value


def near(central_value, half_interval, value):
    """'abs(central_value - value) < half_interval'."""

    return abs(central_value - value) < half_interval


def if_then(condition, value_if_true, value_if_false):
    """Return a conditional value.

    returns -- 'value_if_true' if 'condition' is true; otherwise
    'value_if_false'.  Note that the types of the two possible results
    are coerced together."""

    value_if_true, value_if_false = coerce(value_if_true, value_if_false)

    if condition:
        return value_if_true
    else:
        return value_if_false


def sum(values):
    """Return the sum of 'values'."""

    return reduce(lambda a, b: a + b, values, 0)


def product(values):
    """Return the product of 'values'."""

    return reduce(lambda a, b: a * b, values, 1)


def mean(values):
    """Return the mean of a sequence of values.

    'values' -- A sequence or iterable of numbers."""
    
    length = 0
    sum = 0
    for value in values:
        length += 1
        sum += value
    if length == 0:
        raise ValueError, "no values"
    return sum / length


def variance(values, sample=False):
    """Return the variance of a sequence of values.

    'values' -- A sequence or iterable of numbers."""
    
    num = 0
    sum = 0
    sum_of_squares = 0
    for value in values:
        num += 1
        sum += value
        sum_of_squares += value * value
    if num == 0:
        raise ValueError, "no values"
    mean = sum / num
    result = sum_of_squares / num - mean * mean
    if sample:
        result *= num / (num - 1)
    return result


def standard_deviation(values, sample=False):
    """Return the standard deviation of a sequence of values.

    'values' -- A sequence or iterable of numbers."""

    return math.sqrt(variance(values, sample))


def constrain(value, lo=None, hi=None):
    """Return 'value' constrained to the range '(lo, hi)'.

    If 'lo' or 'hi' is 'None', that constraint is ignored."""

    if lo is not None and value < lo:
        return lo
    if hi is not None and value > hi:
        return hi
    return value


def range(min, max=None, step=1):
    """Return an iterator over a range of values from 'min' to 'max'.

    returns -- An iterator over values starting at 'min' and less than
    'max', each incremented by 'step'."""
    
    # Allow 'range(n)' as shorthand for 'range(0, n)'.
    if max is None:
        max = min
        min = 0

    x = min
    while x < max:
        yield x
        x += step


def multirange(*ranges):
    """Generate the outer product of ranges.

    '*ranges' -- Pairs of the form '(lo, hi)'.

    returns -- A generator over the outer product of the specified
    ranges.  Each generated element is a tuple with number of elements
    equal to the number of arguments ('*ranges')."""

    if len(ranges) > 1:
        for x in range(*ranges[0]):
            for y in multirange(*ranges[1 :]):
                yield (x, ) + y
    else:
        for x in range(*ranges[0]):
            yield (x, )
    

def grid(count, lo, hi):
    x = lo
    step = (hi - lo) / (count - 1)
    for c in xrange(count - 1):
        yield x
        x += step
    yield hi


def multigrid(*ranges):
    if len(ranges) > 1:
        for x in grid(*ranges[0]):
            for y in multigrid(*ranges[1 :]):
                yield (x, ) + y
    else:
        for x in grid(*ranges[0]):
            yield (x, )
   

def intervalUnion(interval0, interval1):
    return (min(interval0[0], interval1[0]),
            max(interval0[1], interval1[1]))


def isIntervalSubset(interval0, interval1):
    """Return true if 'interval0' is a subset of 'interval1'."""
    
    return interval0[0] >= interval1[0] and interval0[1] <= interval1[1]


def dividerange(lo, hi, count):
    for i in xrange(count):
        interval_lo = (hi - lo) * i / count
        interval_hi = (hi - lo) * (i + 1) / count
        yield interval_lo, interval_hi
    

def unweight(samples, weight_fn, random=random.random):
    """Unweight 'samples' by acceptance-rejection.

    Generate an unweighted subset of 'samples'.  The probability of
    including a sample is computed by 'weight_fn'.

    'samples' -- An iterable.

    'weight_fn' -- Function that computes the weight of a sample.  Must
    return a number between zero and one.

    'random' -- A function that generates uniform random numbers between
    zero and one.

    yields -- A subset of 'samples'."""

    for sample in samples:
        weight = weight_fn(sample)
        assert 0 <= weight <= 1
        if random() < weight:
            yield sample


#-----------------------------------------------------------------------
# stuff that depends on CERNLIB
#-----------------------------------------------------------------------

try:
    from hep.cernlib.num import *
except ImportError:
    pass

