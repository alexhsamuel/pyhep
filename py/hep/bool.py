#-----------------------------------------------------------------------
#
# module hep.bool
#
# Copyright 2003 by Alex Samuel.  All rights reserved.
#
#-----------------------------------------------------------------------

"""Implementation of bool for older Python versions.

This module provides 'bool', 'True', and 'False' similar to those found
in recent versions of Python.  These symbols are defined only when
Python doesn't include its own built-in."""

#-----------------------------------------------------------------------
# imports
#-----------------------------------------------------------------------

from   hep.ext import BoolArray

#-----------------------------------------------------------------------

if not hasattr(__builtins__, "bool"):

    False = 0

    True = 1


    class bool(object):

        def __new__(class_, value):
            if value:
                return True
            else:
                return False



