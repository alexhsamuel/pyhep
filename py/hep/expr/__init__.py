#-----------------------------------------------------------------------
#
# module hep.expr
#
# Copyright 2003 by Alex Samuel.  All rights reserved.
#
#-----------------------------------------------------------------------

"""Compiled arithmetical expressions."""

#-----------------------------------------------------------------------
# imports
#-----------------------------------------------------------------------

from   classes import *
import compiler
from   compiler import isCompiledExpression
import expr_parser
from   hep.bool import *
import hep.ext
import inspect
import op
from   op import optimize, substitute, substituteExpressions
from   op import setTypes, setTypesFrom, setTypesFixed
import re
from   symbols import substituteConstants

#-----------------------------------------------------------------------
# functions
#-----------------------------------------------------------------------

def isExpression(object):
    """Return true if 'object' is an expression object."""

    return isCompiledExpression(object) \
           or isinstance(object, Expression)


def asExpression(arg, compile=False, types=None, types_from=None, **names):
    """Interpret 'arg' as an expression.

    For maximum convenience, this function is called by (most)
    user-visible functions which take expression arguments.

    'compile' -- If true, returns a compiled expression.

    '**names' -- Values to substitute for symbol names in the
    expression.

    returns -- If 'arg' is an expression, itself; or 'arg' converted to
    an expression.

    raises -- 'TypeError' if 'arg' is not an expression and cannot be
    converted to one."""

    # If it's already an expression, just use it.
    if isExpression(arg):
        expr = arg

    # If it's a string, parse it.
    elif isinstance(arg, str):
        try:
            expr = parse(arg)
        except expr_parser.ParseError, exception:
            # Make parse exceptions more useful by including the parsed
            # string. 
            raise expr_parser.ParseError, \
                  "%s, while parsing \"%s\"" % (exception, arg)

    # If it's a function, fish out its argument names, make symbols out
    # of these, and wrap it all in a call expression.
    elif callable(arg):
        if inspect.isfunction(arg):
            # An ordinary function.
            symbol_names = inspect.getargspec(arg)[0]
        elif inspect.ismethod(arg):
            # A class method.
            symbol_names = inspect.getargspec(arg.im_func)[0]
            # If it's a bound method, skip the first ('self') argument.
            if arg.im_self is not None:
                symbol_names = symbol_names[1 :]
        elif True:
            symbol_names = ("x", )
        else:
            # Don't know how to get the symbol names.
            raise ValueError, \
                  "don't know how to get argument names for '%r'" \
                  % arg
        # Construct the expression.
        subexprs = map(Symbol, symbol_names)
        expr = Call(Constant(arg), subexprs)

    # If it's numerical, wrap it in a constant.
    elif type(arg) in (int, float, long):
        expr = Constant(arg)

    # Don't know how to handle other things.
    else:
        raise TypeError, "argument must be an expression"

    # Substitute names.
    expr = substitute(expr, **names)
    # Set types.
    if types_from is not None:
        expr = setTypesFrom(expr, **types_from)
    if types is not None:
        expr = setTypes(expr, **types)
    # Compile it, if requested.
    if compile:
        expr = compiler.compile(expr)
    # All done.
    return expr


def parse(expression_string, **names):
    expression = expr_parser.parse(expression_string)
    expression = substitute(expression, **names)
    return expression


def compile(expression, default_type=None, **types):
    """Compile 'expression'.

    'expression' -- """

    if isCompiledExpression(expression):
        # It is already compiled.  
        return expression
    expression = asExpression(expression)
    
    if default_type is not None:
        # Set default symbol types.
        expression = op.setTypesFixed(expression, None, default_type)
    if len(types) != 0:
        # Set specific symbol types.
        expression = op.setTypes(expression, **types)

    # Do the compilation.
    expression = op.optimize(expression)
    return compiler.compile(expression)


def parseFile(lines, symbols={}, expressions=None):
    if expressions is None:
        expressions = {}

    name_regex = re.compile(r"(\w+)\s*=(\s|\Z)")

    name = None
    text = None
    def finish():
        if name is None:
            return
        expression = substituteExpressions(parse(text), expressions)
        expression = substitute(expression, **symbols)
        expressions[name] = expression

    for line in lines:
        line = line.strip()
        # Trim off comments.
        if "#" in line:
            line = line[:line.index("#")]
        # Skip blank lines.
        if line.strip() == "":
            continue

        # Is this the start of a new definition?
        match = name_regex.match(line)
        if match is not None:
            finish()
            name = match.group(1)
            text = line[match.end():]

        elif name is not None:
            text += " " + line

        else:
            raise RuntimeError, "syntax error"
        
    finish()
    return expressions


def _getSymbolNames(expr, result):
    if isinstance(expr, Symbol):
        name = expr.symbol_name
        result[name] = None
    for subexpr in expr.subexprs:
        _getSymbolNames(subexpr, result)


def getSymbolNames(expr):
    """Return a list of names of symbols in 'expr'."""
    
    result = {}
    _getSymbolNames(expr, result)
    return result.keys()
