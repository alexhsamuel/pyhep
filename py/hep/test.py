#-----------------------------------------------------------------------
#
# module hep.test
#
# Copyright 2003 by Alex Samuel.  All rights reserved.
#
#-----------------------------------------------------------------------

"""Regression testing support."""

#-----------------------------------------------------------------------
# imports
#-----------------------------------------------------------------------

from   hep.bool import *
import sys

#-----------------------------------------------------------------------
# exceptions
#-----------------------------------------------------------------------

class TestFailure(Exception):
    pass



#-----------------------------------------------------------------------
# functions
#-----------------------------------------------------------------------

def assert_(expression):
    if not expression:
        raise TestFailure, "assertion failed"


def compare(value, expected, precision=None):
    if type(value) in (list, tuple) and type(expected) in (list, tuple):
        if len(value) != len(expected):
            result = False
        else:
            try:
                result = reduce(
                    lambda r1, r2: r1 and r2,
                    map(lambda (v, e): compare(v, e, precision),
                        zip(value, expected)),
                    True)
            except TestFailure:
                result = False

    elif precision is not None:
        result = (abs(value - expected) <= precision)

    else:
        result = (value == expected)

    if result:
        return True
    else:
        raise TestFailure, \
              "expected %s; got %s" % (repr(expected), repr(value))


def compareSequence(value, expected, ordered=True):
    value = list(value)
    expected = list(expected)
    if not ordered:
        value.sort()
        expected.sort()
    return compare(value, expected)


