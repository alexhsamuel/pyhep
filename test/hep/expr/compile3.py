#-----------------------------------------------------------------------
# imports
#-----------------------------------------------------------------------

from   hep.bool import *
import hep.expr
from   hep.num import *
from   hep.test import compare
from   math import *

#-----------------------------------------------------------------------
# main script
#-----------------------------------------------------------------------

i1 = 42
i2 = -4
f1 = 0.5
f2 = 3.0
symbols = { "i1": i1, "i2": i2, "f1": f1, "f2": f2 }

tests = [
    ("e", e),
    ("pi", pi),
    ("abs(i1)", abs(i1)),
    ("abs(i2)", abs(i2)),
    ("abs(f1)", abs(f1)),
    ("acos(f1)", acos(f1)),
    ("asin(f2 / 4)", asin(f2 / 4)),
    ("atan(i1)", atan(i1)),
    ("atan2(f1, f2)", atan2(f1, f2)),
    ("ceil(f1)", ceil(f1)),
    ("cos(i1)", cos(i1)),
    ("cosh(i1)", cosh(i1)),
    ("exp(i2)", exp(i2)),
    ("floor(f1)", floor(f1)),
    ("gaussian(f1, f2, i2)", gaussian(f1, f2, i2)),
    ("get_bit(i1, 1)", get_bit(i1, 1)),
    ("get_bit(i1, 2)", get_bit(i1, 2)),
    ("hypot(i1, i2)", hep.num.hypot(i1, i2)),
    ("hypot(i1, i2, f1)", hep.num.hypot(i1, i2, f1)),
    ("in_range(i2, 0, i1)", in_range(i2, 0, i1)),
    ("in_range(i2, -6, i1)", in_range(i2, -6, i1)),
    ("in_range(f1, f2, i1)", in_range(f1, f2, i1)),
    ("in_range(i2, f2, f1)", in_range(i2, f2, f1)),
    ("log(i1)", log(i1)),
    ("max(i1, i2)", max(i1, i2)),
    ("max(f1, f2)", max(f1, f2)),
    ("min(i1, i2, 10, 20)", min(i1, i2, 10, 20)),
    ("min(-17.0, f1, f2, 10)", min(-17.0, f1, f2, 10)),
    ("near(40, 3, i1)", near(40, 3, i1)),
    ("near(0, 1, f2)", near(0, 1, f2)),
    ("sin(i1)", sin(i1)),
    ("sinh(i2)", sinh(i2)),
    ("sqrt(f1)", sqrt(f1)),
    ("tan(f2)", tan(f2)),
    ("tanh(f2)", tanh(f2)),
    ]

for expression, expected in tests:
    compiled = hep.expr.asExpression(
        expression, types_from=symbols, compile=True)
    assert hep.expr.isCompiledExpression(compiled)
    print compiled
    
    value = compiled.evaluate(symbols)
    compare(value, expected)
    compare(type(value), type(expected))

