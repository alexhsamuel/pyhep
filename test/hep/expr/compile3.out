2.71828182846
3.14159265359
abs(i1)
abs(i2)
abs(f1)
acos(f1)
asin(f2 / 4)
atan(i1)
atan2(f1, f2)
ceil(f1)
cos(i1)
cosh(i1)
exp(i2)
floor(f1)
gaussian(f1, f2, i2)
get_bit(i1, 1)
get_bit(i1, 2)
hypot(i1, i2)
hypot(i1, i2, f1)
in_range(i2, 0, i1)
in_range(i2, -(6), i1)
in_range(f1, f2, i1)
in_range(i2, f2, f1)
log(i1)
max(i1, i2)
max(f1, f2)
min(i1, i2, 10, 20)
min(-(17.0), f1, f2, 10)
near(40, 3, i1)
near(0, 1, f2)
sin(i1)
sinh(i2)
sqrt(f1)
tan(f2)
tanh(f2)
